client2/thin.go

```diff
@@ -0,0 +1,308 @@
+	if err != nil {
+		return err
+	}
+
+	t.destUnixAddr, err = net.ResolveUnixAddr("unixpacket", "@katzenpost")
+	if err != nil {
+		return err
+	}
+
+	t.log.Debugf("Dial unixpacket %s %s", srcUnixAddr, t.destUnixAddr)
+	t.unixConn, err = net.DialUnix("unixpacket", srcUnixAddr, t.destUnixAddr)
+	if err != nil {
+		return err
+	}
+
+	// WAIT UNTIL we have a Noise cryptographic connection with an edge node
+	t.log.Debugf("Waiting for a connection status message")
+	message1, err := t.readNextMessage()
+	if err != nil {
+		return err
+	}
+	if !message1.IsStatus {
+		panic("did not receive a connection status message")
+	}
+	if !message1.IsConnected {
+		return errors.New("not connected")
+	}
+
+	t.log.Debugf("Waiting for a PKI doc message")
+	message2, err := t.readNextMessage()
+	if err != nil {
+		return err
+	}
+	t.parsePKIDoc(message2.Payload)
+	t.Go(t.worker)
+	t.log.Debug("Dial end")
+	return nil
+}
+
+func (t *ThinClient) worker() {
+	for {
+		select {
+		case <-t.HaltCh():
+			return
+		default:
+		}
+
+		message, err := t.readNextMessage()
+		if err != nil {
+			t.log.Infof("thin client ReceiveMessage failed: %v", err)
+		}
+		if message == nil {
+			return
+		}
+
+		t.log.Debug("THIN CLIENT WORKER RECEIVED A MESSAGE")
+
+		switch {
+		case message.IsStatus == true:
+			t.isConnected = message.IsConnected
+		case message.IsPKIDoc == true:
+			t.parsePKIDoc(message.Payload)
+		default:
+			if message.Payload == nil {
+				t.log.Infof("message.Payload is nil")
+			}
+			response := ThinResponse{
+				SURBID:  message.SURBID,
+				ID:      message.ID,
+				Payload: message.Payload,
+			}
+			select {
+			case <-t.HaltCh():
+				return
+			case t.receivedCh <- response:
+			}
+		}
+	}
+}
+
+func (t *ThinClient) parsePKIDoc(payload []byte) error {
+	doc := &cpki.Document{}
+	err := doc.Deserialize(payload)
+	if err != nil {
+		t.log.Errorf("failed to unmarshal CBOR PKI doc: %s", err.Error())
+		return err
+	}
+	t.pkidocMutex.Lock()
+	t.pkidoc = doc
+	t.pkidocMutex.Unlock()
+	return nil
+}
+
+// PKIDocument returns the thin client's current reference to the PKI doc
+func (t *ThinClient) PKIDocument() *cpki.Document {
+	t.pkidocMutex.RLock()
+	defer t.pkidocMutex.RUnlock()
+	return t.pkidoc
+}
+
+// SendMessageWithoutReply sends a message encapsulated in a Sphinx packet, without any SURB.
+// No reply will be possible.
+func (t *ThinClient) SendMessageWithoutReply(payload []byte, destNode *[32]byte, destQueue []byte) error {
+	req := new(Request)
+	req.WithSURB = false
+	req.IsSendOp = true
+	req.Payload = payload
+	req.DestinationIdHash = destNode
+	req.RecipientQueueID = destQueue
+	req.IsSendOp = true
+
+	blob, err := cbor.Marshal(req)
+	if err != nil {
+		return err
+	}
+	count, _, err := t.unixConn.WriteMsgUnix(blob, nil, t.destUnixAddr)
+	if err != nil {
+		return err
+	}
+	if count != len(blob) {
+		return fmt.Errorf("SendMessage error: wrote %d instead of %d bytes", count, len(blob))
```

Comment:
Maybe this should be a panic; this really shouldn't happen (non-nil err, `count!=len(blob)`) ? I'd expect a `err` at least.